include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -3361,6 +3361,230 @@ namespace internal
+  {
+    const int n_points_1d = n_points_1d_template != -1 ?
+                              n_points_1d_template :
+                              n_points_1d_non_template;
+
+    Assert(n_points_1d > 0, ExcNotImplemented());
+    Assert(n_points_1d < 100, ExcNotImplemented());
+
+    unsigned int compressed_index[100];
+    compressed_index[0] = 0;
+    for (int i = 1; i < n_points_1d - 1; ++i)
+      compressed_index[i] = 1;
+    compressed_index[n_points_1d - 1] = 2;
+
+    // Helper function that extracts the weight. Before doing so,
+    // it checks the values already set.
+    const auto set = [](Number &weight, const Number &data) {
+      if (weight == Number(-1.0) || weight == data)
+        {
+          weight = data;
+          return true; // failure for the entry
+        }
+
+      return false; // success for the entry
+    };
+
+    for (unsigned int c = 0; c < n_components; ++c)
+      for (int k = 0; k < (dim > 2 ? n_points_1d : 1); ++k)
+        for (int j = 0; j < (dim > 1 ? n_points_1d : 1);
+             ++j, data += n_points_1d)
+          {
+            const unsigned int shift =
+              9 * compressed_index[k] + 3 * compressed_index[j];
+
+            if (!set(weights[shift], data[0]))
+              return false; // failure
+
+            for (int i = 1; i < n_points_1d - 1; ++i)
+              if (!set(weights[shift + 1], data[i]))
+                return false; // failure
+
+            if (!set(weights[shift + 2], data[n_points_1d - 1]))
+              return false; // failure
+          }
+
+    return true; // success
+  }
+
+
+  template <int dim,
+            int n_inner_1d_template,
+            int n_extra_1d_template,
+            typename Number>
+  inline void
+  weight_fe_q_dofs_by_entity(const Number *     weights,
+                             const unsigned int n_components,
+                             const int          n_inner_1d_non_template,
+                             const int          n_extra_1d_non_template,
+                             Number *           data)
+  {
+    const int n_inner_1d =
+      n_inner_1d_template != -1 ? n_inner_1d_template : n_inner_1d_non_template;
+
+    const int n_extra_1d =
+      n_extra_1d_template != -1 ? n_extra_1d_template : n_extra_1d_non_template;
+
+    const int n_points_1d = n_inner_1d + 2 * n_extra_1d;
+
+    Assert(n_points_1d > 0, ExcNotImplemented());
+    Assert(n_points_1d < 100, ExcNotImplemented());
+
+    unsigned int compressed_index[100];
+
+    int c = 0;
+    // left neigbor
+    for (int i = 0; i < n_extra_1d; ++i)
+      compressed_index[c++] = 0;
+
+    // left face
+    compressed_index[c++] = 1;
+
+    // inner part
+    for (int i = 0; i < n_inner_1d - 2; ++i)
+      compressed_index[c++] = 2;
+
+    // right face
+    compressed_index[c++] = 3;
+
+    // right neighbor
+    for (int i = 0; i < n_extra_1d; ++i)
+      compressed_index[c++] = 4;
+
+    AssertDimension(c, Utilities::pow(5, dim));
```

Comment:
Did you check this assertion? To me, it seems that you were only running through a 1D loop at this point, so the value would not be 5 but `n_inner_1d + 2*n_extra_1d = n_points_1d`?